core/src/main/java/org/apache/druid/math/expr/CartesianList.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.math.expr;
+
+import com.google.common.base.Preconditions;
+import com.google.common.math.IntMath;
+
+import javax.annotation.Nullable;
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.RandomAccess;
+
+/**
+ * {@link CartesianList} computes the cartesian product of n lists. It is adapted from and is *nearly* identical to one
+ * Guava CartesianList which comes from a version from "the future" that we don't yet have, with the key difference that
+ * it is not {@link com.google.common.collect.ImmutableList} based, so it can hold null values to be compatible with the
+ * evaluation and handling of cartesian products of expression arrays with null elements, e.g. ['a', 'b', null]
+  */
+
+public final class CartesianList<E> extends AbstractList<List<E>> implements RandomAccess
```

Comment:
skipping this as this is just a clone from guava.
nit: it would be nice to create a separate package in "core/src/main/java/org/apache/druid/guava/..." for all files copied from guava to track them better.

Reply:
Heh, I guess we _do_ have a guava package, `core/src/main/java/org/apache/druid/common/guava`:
![Screen Shot 2019-06-03 at 1 42 52 PM](https://user-images.githubusercontent.com/1577461/58833386-c46dcf00-8605-11e9-8996-7024543618b4.png)

but amusingly _none_ of the files listed in `LICENSE` as being modified from guava are in that folder:
![Screen Shot 2019-06-03 at 1 46 25 PM](https://user-images.githubusercontent.com/1577461/58833494-0139c600-8606-11e9-9c4e-e871e78554a1.png)